### PR TITLE
Device Registration changes for a MAM resource

### PR DESCRIPTION
**Pull request by kreedula**
- Added new exception to track this specific scenerio.
- Also changed MSAL version to 0.3.0

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/ExceptionAdapter.java [hunk 1]
```diff
@@ -63,6 +65,22 @@ public static BaseException exceptionFromAcquireTokenResult(final AcquireTokenRe
...
+                            MicrosoftAuthorizationErrorResponse microsoftAuthorizationErrorResponse =
+                                    (MicrosoftAuthorizationErrorResponse) authorizationErrorResponse;
+
+                            if(microsoftAuthorizationErrorResponse.getError().equals(
+                                    MicrosoftAuthorizationErrorResponse.DEVICE_REGISTRATION_NEEDED)){
+
+                                return new DeviceRegistrationRequiredException(
+                                        microsoftAuthorizationErrorResponse.getError(),
+                                        microsoftAuthorizationErrorResponse.getErrorDescription()
+                                        , microsoftAuthorizationErrorResponse.getUserName()
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/ExceptionAdapter.java]: Nit: Just some strange formatting here

File: common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java [hunk 2]
```diff
@@ -65,7 +65,7 @@
 public class AzureActiveDirectoryWebViewClient extends OAuth2WebViewClient {
     private static final String TAG = AzureActiveDirectoryWebViewClient.class.getSimpleName();
     private static final String INSTALL_URL_KEY = "app_link";
-    private static final String INSTALL_UPN_KEY = "username";
+    public static final String INSTALL_UPN_KEY = "username";
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java]: Should we consider graduating these constants out of this file? Not necessarily that it doesn't belong here (it's set here, after all) but, I notice that now the `MicrosoftStsAuthorizationResultFactory` is reaching-into this class to grab this constant which is maybe a little weird organizationally speaking.

File: common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java [hunk 3]
```diff
@@ -93,11 +113,18 @@ public MicrosoftStsAuthorizationResult createAuthorizationResult(final int resul
                 result = createAuthorizationResultWithErrorResponse(AuthorizationStatus.FAIL,
                         MicrosoftAuthorizationErrorResponse.DEVICE_REGISTRATION_NEEDED,
                         MicrosoftAuthorizationErrorResponse.DEVICE_REGISTRATION_NEEDED);
+                // Set username returned from the service
+                result.getAuthorizationErrorResponse().setUserName(data.getStringExtra(
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java]: As one minor additional note, something here looks 'unsafe' -- will the error response always be here? will it always have a username? Is there no need at all to check that these values `!= null` ?

Chalk my questions up to paranoia 😅

Comment (kreedula) [reply common/src/main/java/com/microsoft/identity/common/internal/providers/microsoft/microsoftsts/MicrosoftStsAuthorizationResultFactory.java]: Thanks for the raising this. I have double checked, `
createAuthorizationResultWithErrorResponse` will return you a non-null error response, so we are good here.

File: common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java (hunk 2 above)
Comment (kreedula) [reply common/src/main/java/com/microsoft/identity/common/internal/ui/webview/AzureActiveDirectoryWebViewClient.java]: Good call, I was being lazy here and just changed the constant to public. Moved to `AuthenticationConstants` as appropriate

**Pull request merged**